protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+package protofsm
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+const (
+	// pollInterval is the interval at which we'll poll the SendWhen
+	// predicate if specified.
+	pollInterval = time.Millisecond * 100
+)
+
+// EmitedEvent is a special type that can be emitted by a state transition.
+// This can container internal events which are to be routed back to the state,
+// or external events which are to be sent to the daemon.
+type EmittedEvent[Event any] struct {
+	// InternalEvent is an optional internal event that is to be routed
+	// back to the target state. This enables state to trigger one or many
+	// state transitions without a new external event.
+	InternalEvent fn.Option[Event]
+
+	// ExternalEvent is an optional external event that is to be sent to
+	// the daemon for dispatch. Usually, this is some form of I/O.
+	ExternalEvents fn.Option[DaemonEventSet]
```

Comment:
Is there a material difference in semantics between `None` and `Some({})` here?

Reply:
I think setting `Some({})` would be considered a bug, or quirky logic. This should only be set to `Some` when the FSM has some I/O that it wants to perform. Similar to the whole `nil` vs `[]byte{}` thing with slices.

Reply:
Can we just make it a set then, since sets can be empty?